Shift adding PodScheduled from kube-apiserver to kube-scheduler

rohitsakala opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR fixes a bug where the PodScheduled condition was being created by kube-apiserver. When the scheduler fails to schedule a Pod, the kube-scheduler creates the [PodScheduled condition](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/schedule_one.go#L1100). Logically, using the same argument, kube-scheduler should create the PodScheduled condition.

Another argument on why kube-scheduler should create this condition is that the kube-scheduler assigns the node to the pod and not the kube-apiserver.

In addition to the above correction, it fixes https://github.com/kubernetes/kubernetes/issues/127508. 

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/127508

#### Does this PR introduce a user-facing change?
No

```release-note
PodScheduled condition is now created by kube-scheduler instead of kube-apiserver
```

I am new contributor. Please suggest me the right solution if this is not the correct solution.

AxeZhan commented on pkg/scheduler/schedule_one.go:
```diff
@@ -129,6 +129,15 @@ func (sched *Scheduler) ScheduleOne(ctx context.Context) {
+
+		// Usually, DonePod is called inside the scheduling queue,
+		// but in this case, we need to call it here because this Pod won't go back to the scheduling queue.
+		sched.SchedulingQueue.Done(assumedPodInfo.Pod.UID)
```
Is this change relevant to the issue?
I thought `Done` should already be called inside the binding cycle if pod passes `permit` and `prebind`, no?

AxeZhan commented on pkg/scheduler/schedule_one.go:
```diff
@@ -129,6 +129,11 @@ func (sched *Scheduler) ScheduleOne(ctx context.Context) {
 			return
 		}
+		// Update the PodCondition condition to reflect that the pod is scheduled.
+		podutil.UpdatePodCondition(&pod.Status, &v1.PodCondition{
```
Can you add some checks for this in tests to ensure the condition is updated?
For example, we can add some check to see if pod condition is updated after scheduling in `TestSchedulerScheduleOne`.

AxeZhan replied:
Okay, I went back and checked the issue.
Me myself feel reasonable to let scheduler set the `PodScheduled` Condition. But this is not mentioned in the issue. The issue simply wants this condition to be owned.
In this case, I think we'll need to wait for sig-node folks to see if it's ok with them.

BTW, can we let the condition be owned by the kubelet, but still allow scheduler to set it to false?(I'm not familiar with this part). Even if we can do this, it feels a little weird.

AxeZhan replied:
Seems we both feel logical to let the scheduler own the condition.
Then can you add some tests for this?
I think we'll need a UT to make sure condition is updated after ScheduleOne.(Maybe just add a check in existing tests like I said before).
Any maybe an e2e test to check the `managedFields` after pod scheduled?

rohitsakala committed:
Add unit test for PodScheduled condition

unit

macsko commented on pkg/scheduler/schedule_one.go:
```diff
@@ -309,12 +309,21 @@ func (sched *Scheduler) bindingCycle(
 	// We can call Done() here because
 	// we can free the cluster events stored in the scheduling queue sonner, which is worth for busy clusters memory consumption wise.
 	sched.SchedulingQueue.Done(assumedPod.UID)
+	podScheduled := &v1.PodCondition{
+		Type:   v1.PodScheduled,
+		Status: v1.ConditionTrue,
+	}
+	newStatus := assumedPod.Status.DeepCopy()
+	podutil.UpdatePodCondition(newStatus, podScheduled)
+	if err := util.PatchPodStatus(ctx, sched.client, assumedPod, newStatus); err != nil {
+		logger.Error(err, "Error updating pod with PodScheduled condition", "pod", klog.KObj(assumedPod))
+	}
```
Shouldn't the condition be set after binding?

macsko replied:
> `Any failures after this point cannot lead to the Pod being considered unschedulable` so I added it there.

Unschedulable != unscheduled in this context. This comment says that after this point, if the binding fails, pod can't be considered unschedulable, but if an error happens it is treated as scheduling error. In this case, pod is still unscheduled as it will need to go through scheduling cycle again. It means that status might be inaccurate, but fortunately will be updated to failure in `handleSchedulingFailure` ([ref](https://github.com/kubernetes/kubernetes/blob/a499facee693a1a83daadb82d88f7b51d324ffc5/pkg/scheduler/schedule_one.go#L1100-L1105)).

haosdent commented on pkg/scheduler/schedule_one.go:
```diff
@@ -309,12 +309,21 @@ func (sched *Scheduler) bindingCycle(
+		Status: v1.ConditionTrue,
+	}
+	newStatus := assumedPod.Status.DeepCopy()
+	podutil.UpdatePodCondition(newStatus, podScheduled)
```
I think you need to check the update result at here? Similiar to https://github.com/kubernetes/kubernetes/blob/1504f10e7946f95a8b1da35e28e4c7453ff62775/pkg/scheduler/framework/preemption/preemption.go#L181-L187

haosdent commented on pkg/scheduler/schedule_one.go:
```diff
@@ -309,12 +309,21 @@ func (sched *Scheduler) bindingCycle(
+	}
 
 	// Run "bind" plugins.
 	if status := sched.bind(ctx, fwk, assumedPod, scheduleResult.SuggestedHost, state); !status.IsSuccess() {
```
Seems `assumedPod`'s status is still the old one because https://github.com/kubernetes/kubernetes/blob/1504f10e7946f95a8b1da35e28e4c7453ff62775/pkg/scheduler/util/utils.go#L105-L134 didn't modify `old.Status` (or it would modify?)

I am not sure if `assumePod` keep the old status match your expecation at here.

rohitsakala committed:
Remove setting PodScheduled condition from kubelet

AxeZhan commented:
/ok-to-test

rohitsakala committed:
Shift adding PodScheduled from kube-apiserver to kube-scheduler

Adding a PodScheduled when a failure occurs is done in kube-scheduler. Naturally adding a PodScheduled condition when a success occurs should be done in the
kube-scheduler.

AxeZhan commented on pkg/scheduler/schedule_one_test.go:
```diff
@@ -852,6 +868,17 @@ func TestSchedulerScheduleOne(t *testing.T) {
 				if diff := cmp.Diff(item.expectBind, gotBinding); diff != "" {
 					t.Errorf("got binding diff (-want, +got): %s", diff)
 				}
+				if item.eventReason == "Scheduled" {
```
Instead of adding a new wrapping function `podWithCondition`, and update a lot manually in L846-L855.
I think we can just simply iterate pod's conditions when `item.eventReason == "Scheduled"` and check if the correct condition `{ Type:   v1.PodScheduled, Status: v1.ConditionTrue}` is contained?

Also, can we also add a check for `managedFields`(see if it's set correctly) as this is the origin purpose of the issue.

AxeZhan replied:
Ok, got it, So we actually have already compared the condition.
Then yes, we should keep the `podWithCondition`, but still we don't have to modify the `LastTransitionTime` manually, instead we can just ignore them by using `cmp.DIff` instead of `reflect.DeepEqual`

```
	if diff := cmp.Diff(pod2, pod1, cmpopts.IgnoreFields(v1.PodCondition{}, "LastTransitionTime")); diff != "" {
		fmt.Println(diff)
	}
```

> Also, can you point me to a place where managedFields testing is done scheduler ?

I'm also not familiar with this. Will take a look later.

AxeZhan replied:
> @AxeZhan I have modified an existing e2e test to test the managedFields condition.

That's great, can we use `cmp.Diff` here to remove the modify to `LastTransitionTime` though?

rohitsakala committed:
Modify existing e2e test to test PodScheduled condition

AxeZhan commented on test/e2e/scheduling/priorities.go:
```diff
@@ -255,6 +255,17 @@ var _ = SIGDescribe("SchedulerPriorities", framework.WithSerial(), func() {
+		managedFields := tolePod.GetManagedFields()
+		PodScheduledConditionFound := false
+		for _, entry := range managedFields {
+			if entry.FieldsV1.String() == `{"f:status":{"f:conditions":{".":{},"k:{\"type\":\"PodScheduled\"}":{".":{},"f:lastProbeTime":{},"f:lastTransitionTime":{},"f:status":{},"f:type":{}}}}}` {
```
This may lead to subsequent minor changes causing the tests to fail.

I'm thinking maybe a simple `if strings.Contains(entry.FieldsV1.String(),"PodScheduled")` is enough?

AxeZhan commented on pkg/scheduler/schedule_one_test.go:
```diff
@@ -837,21 +843,30 @@ func TestSchedulerScheduleOne(t *testing.T) {
+					if err != nil {
+						t.Fatal(err)
+					}
+					if diff := cmp.Diff(updatedPod.Status.Conditions, podStatusConditions, cmpopts.IgnoreFields(v1.PodCondition{}, "LastTransitionTime")); diff != "" {
```
```suggestion
					if diff := cmp.Diff(podStatusConditions, updatedPod.Status.Conditions, cmpopts.IgnoreFields(v1.PodCondition{}, "LastTransitionTime")); diff != "" {
					t.Errorf("unexpected conditions (-want, +got):\n%s", diff)
```

Print diff when there is. Also do this for above lines.

AxeZhan reviewed (commented):
Thanks, lgtm from scheduling side. Only one nit.

AxeZhan commented:
/lgtm

/cc @sergeykanzhelev @kerthcet  
As robot suggests.

haircommander commented:
/lgtm

@SergeyKanzhelev @mrunalp for approval

SergeyKanzhelev commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -675,9 +675,6 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 	updateLastTransitionTime(&status, &oldStatus, v1.PodReadyToStartContainers)
 
-	// Set PodScheduledCondition.LastTransitionTime.
-	updateLastTransitionTime(&status, &oldStatus, v1.PodScheduled)
```
do we need to keep it as a safeguard? Or can it cause troubles? Sorry, just glancing over the PR for now.

SergeyKanzhelev commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -675,9 +675,6 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 	// Set PodReadyToStartContainersCondition.LastTransitionTime.
 	updateLastTransitionTime(&status, &oldStatus, v1.PodReadyToStartContainers)
 
-	// Set PodScheduledCondition.LastTransitionTime.
```
so this will not be a problem for a version skew as the api version will be at least the same as kubelet version so will have the fix available. Is there any other weird cases of a version skew to worry about?

kerthcet commented on pkg/scheduler/schedule_one.go:
```diff
@@ -309,12 +309,25 @@ func (sched *Scheduler) bindingCycle(
+	newStatus := assumedPod.Status.DeepCopy()
+	updated := podutil.UpdatePodCondition(newStatus, podScheduled)
+	if updated {
+		if err := util.PatchPodStatus(ctx, sched.client, assumedPod, newStatus); err != nil {
```
If we haven't bind the pod, how can we say: hey, let's update the status to scheduled? I think it doesn't make sense.

And I think the complexity sits at `If we bind successfully, but update the status in failure, who will be responsible for populating the status again? I think for scheduler, if bind successfully, we're end now. 

Maybe delegate the condition update to kubelet is not a bad idea but it should be put in the first place?

kerthcet replied:
But have no idea whether this will lead to unexpected behaviors like depending on the podScheduled condition, I think we only give the promise that **Pod is scheduled once spec.nodeName is not empty**.

Huang-Wei replied:
@kerthcet 's concern is valid I think. It's because other than default-scheduler, we can have other players that take over a pod's scheduling lifecycle:

- a pod can be created with a non-nil `.spec.nodeName` - in this case, (default-)scheduler isn't involved at all
- a custom scheduler (yunikorn/volcano/etc.) - they have their own logic from the scratch

Huang-Wei replied:
> This PR fixes a bug where the PodScheduled condition was being created by kube-apiserver.

I don't think it's a _bug_ actually.

> Logically, using the same argument, kube-scheduler should create the PodScheduled condition.

I'd say kube-apiserver, scheduler or scheduler's variant, kubelet can jointly maintain the condition. It's doesn't make to declare kube-scheduler should create (or own) the PodScheduled condition.

> What would you suggest now ? How do I go forward?

I'd say this PR is not on the right direction. We should tackle #127508 with merely fixing the right manager field.

Huang-Wei replied:
> What do you think which component should set it ?

I'm not quite sure if the original issue is general to reproduce, or need some particular sequences of steps. But generally, the manager field should be non-empty and set by the component who updates the PodScheduled condition.

tallclair replied:
If a pod is bound to a node, but the Kubelet is unresponsive, should the condition be set? Semantically it seems like the answer should be yes, which means it needs to be set by the API server.

If the condition is set by the API server, it should also check for `nodeName` being set on pod creation, and add the condition in that case.

vahan-sahakyan-op commented:
/approve
